chainer/link_hooks/spectral_normalization.py

```diff
@@ -0,0 +1,262 @@
+from chainer import backend
+from chainer import configuration
+import chainer.functions as F
+import chainer.links as L
+from chainer import link_hook
+from chainer import variable
+
+
+def l2normalize(xp, v, eps=1e-12):
+    """Normalize a vector by its L2 norm.
+
+    Args:
+        xp (numpy or cupy):
+        v (numpy.ndarray or cupy.ndarray)
+        eps (float): Epsilon value for numerical stability.
+
+    Returns:
+        :class:`numpy.ndarray` or :class:`cupy.ndarray`
+
+    """
+    return v / (xp.linalg.norm(v) + eps)
+
+
+def update_approximate_vectors(
+        weight_matrix, u, n_power_iteration=1, eps=1e-12):
+    """Update the first left and right singular vectors.
+
+    This function updates the first left singular vector `u` and
+    the first right singular vector `v`.
+
+    Args:
+        weight_matrix (~chainer.Variable): 2D weight.
+        u (numpy.ndarray, cupy.ndarray, or None):
+            Vector that approximates the first left singular vector and
+            has the shape of (out_size,).
+        n_power_iteration (int): Number of iterations to approximate
+            the first right and left singular vectors.
+
+    Returns:
+        :class:`numpy.ndarray` or `cupy.ndarray`:
+            Approximate first left singular vector.
+        :class:`numpy.ndarray` or `cupy.ndarray`:
+            Approximate first right singular vector.
+
+    """
+    weight_matrix = weight_matrix.array
+    xp = backend.get_array_module(weight_matrix)
+    for _ in range(n_power_iteration):
+        v = l2normalize(xp, xp.dot(u, weight_matrix), eps)
+        u = l2normalize(xp, xp.dot(weight_matrix, v), eps)
+    return u, v
+
+
+def calculate_max_singular_value(weight_matrix, u, v):
+    """Calculate max singular value by power iteration method.
+
+    Args:
+        weight_matrix (~chainer.Variable)
+        u (numpy.ndarray or cupy.ndarray)
+        v (numpy.ndarray or cupy.ndarray)
+
+    Returns:
+        ~chainer.Variable: Max singular value via power iteration method.
+
+    """
+    sigma = F.matmul(F.matmul(u, weight_matrix), v)
+    return sigma
+
+
+class SpectralNormalization(link_hook.LinkHook):
+    r"""Spectral Normalization link hook implementation.
+
+    This hook normalizes a weight using max singular value and this value
+    is computed via power iteration method. Currently, this hook is supposed to
+    be added to :func:`chainer.links.Linear`, :func:`chainer.links.EmbedID`,
+    :func:`chainer.links.Convolution2D`, :func:`chainer.links.ConvolutionND`,
+    :func:`chainer.links.Deconvolution2D`,
+    and :func:`chainer.links.DeconvolutionND`. However, you can use this to
+    other links like RNNs by specifying ``weight_name``.
+
+    .. math::
+
+       \mathbf{W} &:=& \dfrac{\mathbf{W}}{\sigma(\mathbf{W})} \\
```

Comment:
Since `:=` in the next line means "definition", `:=` in this line cannot mean "substitution".  Could you change it to something like `\bar{\mathbf{W}} &:=&`?